### ISSUE TITLE
[tune] trainable:stop not called at end of training

### ISSUE THREAD

**Issue by bask0** [ray-project/ray#5767]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Ubuntu 16
- **Ray installed from (source or binary)**:
From source
- **Ray version**:
0.8.0.dev4
- **Python version**:
3.6

### Describe the problem

I call `ray.tune.run(...)` to optimize 5 cross-validation folds in parallel (the best hyperparameters have already been found in a previous step). For test purposes, I stop the training after 2 epochs. At this point, the `trainable._stop()` should be called to save the final model predictions:

```
def _stop(self):
    # Dump predictions to file.
```

The problem: For the first fold that finishes, `trainable.stop()` is never called. I changed the source code of `ray.tune.trainable:stop()` to print something when the method is called. The first fold that finishes never calls `stop()` and the final model prediction file is missing, even if no error occurs:

```
TERMINATED trials:
 - ModelTrainerMT_0_fold=0:     TERMINATED, [5 CPUs, 0.5 GPUs], [pid=11897], 74 s, 2 iter
 - ModelTrainerMT_1_fold=1:     TERMINATED, [5 CPUs, 0.5 GPUs], [pid=11901], 70 s, 2 iter
 - ModelTrainerMT_2_fold=2:     TERMINATED, [5 CPUs, 0.5 GPUs], [pid=11899], 70 s, 2 iter
 - ModelTrainerMT_3_fold=3:     TERMINATED, [5 CPUs, 0.5 GPUs], [pid=11888], 75 s, 2 iter
 - ModelTrainerMT_4_fold=4:     TERMINATED, [5 CPUs, 0.5 GPUs], [pid=11925], 70 s, 2 iter
```

I appreciate any help or hints, let me know if you need more info.

---

This is the scheduler I use (even though training is stopped after 2 epochs):

EDIT: Maybe not so relevant but I just figured out that using Median Stopping Rule does not make much sense here.

```
early_stopping_scheduler = MedianStoppingRule(
    time_attr='epoch',
    metric=config['metric'],
    mode='min',
    grace_period=0 if args.predict else 40,
)
```

This is how I call `tune.run`:
```
if args.predict:
    best_config = load_best_config(store_train)
    best_config.update({"fold": tune.grid_search([0, 1, 2, 3, 4])})
    tune.run(
        ModelTrainerMT,
        resources_per_trial={
            'cpu': 5,
            'gpu': 0.5,
        },
        num_samples=1,
        config=best_config,
        local_dir=ModelTrainerMT.store,
        raise_on_failed_trial=True,
        verbose=1,
        with_server=False,
        ray_auto_init=False,
        scheduler=early_stopping_scheduler,
        loggers=[JsonLogger, CSVLogger],
        checkpoint_at_end=True,
        reuse_actors=True,
        stop={'epoch': 2 if args.test else config['max_t']} # <--- Training stops after 2 epochs.
    )
```

**Comment by richardliaw** [comment]
Ah; can you try `reuse_actors=False`? This is a bug; I'll push a fix now. Thanks!

**Comment by bask0** [comment]
This fixed the problem for me, thank you!

I leave it up to you to decide whether you want to close this now or after having fixed the bug.
